src/compiler/checker.ts

```diff
@@ -17775,7 +17775,7 @@ namespace ts {
                 target.flags & TypeFlags.Union ? some((target as UnionType).types, t => isTypeDerivedFrom(source, t)) :
                 source.flags & TypeFlags.InstantiableNonPrimitive ? isTypeDerivedFrom(getBaseConstraintOfType(source) || unknownType, target) :
                 target === globalObjectType ? !!(source.flags & (TypeFlags.Object | TypeFlags.NonPrimitive)) :
-                target === globalFunctionType ? !!(source.flags & TypeFlags.Object) && isFunctionObjectType(source as ObjectType) :
+                target === globalFunctionType ? isFunctionObjectType(source) :

```

Comment:
The issue isn't just with `x instanceof Function` checks, it also extends to `x instanceof Object` checks. I have a more general solution to the problem in #51631.